pontoon/batch/views.py

```diff
@@ -0,0 +1,174 @@
+        'fuzzy_strings',
+        'translated_strings',
+    ])
+
+    project = entity.resource.project
+    project.aggregate_stats()
+    locale.aggregate_stats()
+    ProjectLocale.objects.get(locale=locale, project=project).aggregate_stats()
+
+
+def mark_changed_translation(changed_entities, locale):
+    """Mark entities as changed, for later sync.
+    """
+    changed_entities_array = []
+    existing = (
+        ChangedEntityLocale.objects
+        .values_list('entity', 'locale')
+        .distinct()
+    )
+    for changed_entity in changed_entities:
+        key = (changed_entity.pk, locale.pk)
+
+        # Remove duplicate changes to prevent unique constraint violation.
+        if key not in existing:
+            changed_entities_array.append(
+                ChangedEntityLocale(entity=changed_entity, locale=locale)
+            )
+
+    ChangedEntityLocale.objects.bulk_create(changed_entities_array)
+
+
+def update_translation_memory(changed_translation_pks, project, locale):
+    """Update translation memory for a list of translations.
+    """
+    memory_entries = [
+        TranslationMemoryEntry(
+            source=t.entity.string,
+            target=t.string,
+            locale=locale,
+            entity=t.entity,
+            translation=t,
+            project=project,
+        ) for t in (
+            Translation.objects
+            .filter(pk__in=changed_translation_pks)
+            .prefetch_related('entity__resource')
+        )
+    ]
+    TranslationMemoryEntry.objects.bulk_create(memory_entries)
+
+
+@login_required(redirect_field_name='', login_url='/403')
+@require_POST
+@require_AJAX
+@transaction.atomic
+def batch_edit_translations(request):
+    """Perform an action on a list of translations.
+
+    Available actions are defined in `ACTIONS_FN_MAP`. Arguments to this view
+    are defined in `models.BatchActionsForm`.
+
+    """
+    form = forms.BatchActionsForm(request.POST)
+    if not form.is_valid():
+        return HttpResponseBadRequest(form.errors.as_json())
+
+    locale = get_object_or_404(Locale, code=form.cleaned_data['locale'])
+
+    entities = (
+        Entity.objects
+        .filter(pk__in=form.cleaned_data['entities'])
+        .prefetch_translations(locale)
+    )
+
+    if not entities.exists():
+        return JsonResponse({'count': 0})
+
+    # Batch editing is only available to translators. Check if user has
+    # translate permissions for all of the projects in passed entities.
+    projects_pk = entities.values_list('resource__project__pk', flat=True)
+    projects = Project.objects.filter(pk__in=projects_pk.distinct())
+    for project in projects:
+        if not request.user.can_translate(project=project, locale=locale):
+            return HttpResponseForbidden(
+                "Forbidden: You don't have permission for batch editing"
+            )
+
+    translation_pks = set()
+
+    # Find all impacted translations, including plural forms.
```

Comment:
This only collects "active" translations. Impacted can also be non-active translations (in a sense that they get rejected for example).